File: lib/Klib.jl
module Klib

export getopt, GzFile, close, Bufio, readbyte, readuntil!, FastxReader, FastxRecord

#
# Getopt iterator
#
struct Getopt
	args::Array{String}
	ostr::String
	longopts::Array{String}
end

function Base.iterate(g::Getopt, (pos, ind) = (1, 1))
	if g.ostr[1] != '+' # allow options to appear after main arguments
		while ind <= length(g.args) && (g.args[ind][1] != '-' || g.args[ind] == "-")
			ind += 1
		end
	end
	if ind > length(g.args) || g.args[ind][1] != '-' || g.args[ind] == "-" return nothing end
	if length(g.args[ind]) >= 2 && g.args[ind][1] == '-' && g.args[ind][2] == '-'
		if length(g.args[ind]) == 2 # actually, Julia will always filter out "--" in ARGS. Ugh!
			deleteat!(g.args, ind)
			return nothing
		end
		optopt, optarg, pos = "?", "", 0
		if length(g.longopts) > 0
			eqpos = findfirst(isequal('='), g.args[ind])
			a = eqpos == nothing ? g.args[ind][3:end] : g.args[ind][3:eqpos-1]
			n_matches, match = 0, ""
			for l in g.longopts
				r = findfirst(a, l)
				if r != nothing && collect(r)[1] == 1
					n_matches += 1
					match = l
				end
			end
			if n_matches == 1
				optopt = string("--", match[end] == '=' ? match[1:end-1] : match);
				if eqpos != nothing
					optarg = g.args[ind][eqpos+1:end]
				elseif match[end] == '=' && ind + 1 <= length(g.args)
					deleteat!(g.args, ind)
					optarg = g.args[ind]
				end
			end
		end
	else
		if pos == 1 pos = 2 end
		optopt, optarg = g.args[ind][pos], ""
		pos += 1
		i = findfirst(isequal(optopt), g.ostr)
		if i == nothing # unknown option
			optopt = '?'
		elseif i < length(g.ostr) && g.ostr[i + 1] == ':' # require argument
			if pos <= length(g.args[ind])
				optarg = g.args[ind][pos:end]
			elseif ind + 1 <= length(g.args)
				deleteat!(g.args, ind)
				optarg = g.args[ind]
			end
			pos = 0
		end
		optopt = optopt == '?' ? "?" : string('-', optopt)
	end
	if pos == 0 || pos > length(g.args[ind])
		deleteat!(g.args, ind) # FIXME: can be slow when ostr[1] == '-'
		pos = 1
	end
	return ((optopt, optarg), (pos, ind))
end

"""
    getopt(args::Array{String}, ostr::String, longopts::Array{String}=String[])

Iterate through command line options with a getopt-like interface and remove
options from `args`.

`args` is typically `ARGS`. By default, options are allowed to occur after
non-option arguments. If `ostr[1]=='+'`, the default behavior is disabled.

# Examples
```julia
for (opt, arg) in Klib.getopt(ARGS, "xy:", ["foo", "bar="])
	@show (opt, arg)
end
@show ARGS # only non-option arguments remain
```
"""
getopt(args::Array{String}, ostr::String, longopts::Array{String} = String[]) = Getopt(args, ostr, longopts)

#
# ByteBuffer
#
mutable struct ByteBuffer <: AbstractVector{UInt8}
	a::Ptr{UInt8}
	m::UInt64

	function ByteBuffer()
		x = new(C_NULL, 0)
		finalizer(destroy!, x)
		return x
	end
end

function reserve!(b::ByteBuffer, z::UInt64)
	if z > b.m
		x = z - 1; x |= x >> 1; x |= x >> 2; x |= x >> 4; x |= x >> 8; x |= x >> 16; x |= x >> 32; x += 1
		b.m = (x<<1) + (x<<2) >= z ? (x<<1) + (x<<2) : x
		b.a = ccall(:realloc, Ptr{Cvoid}, (Ptr{Cvoid}, Csize_t), b.a, b.m)
	end
end

function destroy!(b::ByteBuffer)
	ret = ccall(:free, Cint, (Ptr{Cvoid},), b.a)
	b.a, b.m = C_NULL, 0
	return ret
end

tostring(b::ByteBuffer, n::Int) = String(unsafe_wrap(Vector{UInt8}, b.a, n))

#
# GzFile
#
mutable struct GzFile <: IO
	fp::Ptr{Cvoid}

	function GzFile(fn::String, mode = "r")
		x = ccall((:gzopen, "libz"), Ptr{Cvoid}, (Cstring, Cstring), fn, mode)
		y = x == C_NULL ? nothing : new(x)
		if y != nothing finalizer(Base.close, y) end
		return y
	end
end

Base.readbytes!(fp::GzFile, buf::Vector{UInt8}) = ccall((:gzread, "libz"), Cint, (Ptr{Cvoid}, Ptr{Cvoid}, Cuint), fp.fp, buf, length(buf))

function Base.close(fp::GzFile)
	ret = fp.fp != C_NULL ? ccall((:gzclose, "libz"), Cint, (Ptr{Cvoid},), fp.fp) : -1
	fp.fp = C_NULL
	return ret
end

#
# Bufio
#
mutable struct Bufio{T<:IO} <: IO # TODO: support writing
	fp::T
	start::Int
	len::Int
	iseof::Bool
	buf::Vector{UInt8}
	bb::ByteBuffer

	Bufio{T}(fp::T, mode::String = "r", bufsize = 0x10000) where {T<:IO} = new{T}(fp, 1, 0, false, Vector{UInt8}(undef, bufsize), ByteBuffer())
end

"""
    Bufio(fp::T, mode::String = "r", bufsize = 0x10000) where T<:IO

Wrap an IO stream and return a `Bufio{T}` object. For the moment, only reading
is supported. `mode` is ignored.
"""
Bufio(fp::T, mode::String = "r", bufsize = 0x10000) where {T<:IO} = Bufio{T}(fp, mode, bufsize)

function Base.readbytes!(fp::Bufio{T}, buf::Vector{UInt8}, len=length(buf)) where {T<:IO}
	if fp.iseof && fp.start > fp.len return 0 end # normal EOF
	offset = 1
	while len > fp.len - (fp.start - 1)
		l = fp.len - (fp.start - 1)
		if length(buf) < offset - 1 + l resize(buf, offset - 1 + l) end
		@inbounds copyto!(buf, offset, fp.buf, fp.start, l)
		len -= l
		offset += l
		fp.start, fp.len = 1, Base.readbytes!(fp.fp, fp.buf)
		if fp.len < length(fp.buf) fp.iseof = true end
		if fp.len == 0 return offset - 1 end
	end
	if length(buf) < offset - 1 + len resize(buf, offset - 1 + len) end
	@inbounds copyto!(buf, offset, fp.buf, fp.start, len)
	fp.start += len
	return offset - 1 + len
end

Base.eof(r::Bufio{T}) where {T<:IO} = r.iseof && r.start > r.len ? true : false

"""
    readbyte(r::Bufio{T}) where T<:IO

Read a byte from `r` and return it on success, or -1 on end-of-file (EOF).

This function distincts from `Base.read(io::IO, UInt8)` in that the latter
throws an exception on EOF.
"""
function readbyte(r::Bufio{T}) where {T<:IO}
	if r.iseof && r.start > r.len return -1 end
	if r.start > r.len
		r.start, r.len = 1, Base.readbytes!(r.fp, r.buf)
		if r.len < length(r.buf) r.iseof = true end
		if r.len == 0 return -1 end
	end
	c = r.buf[r.start]
	r.start += 1
	return c
end

trimret(buf::ByteBuffer, n) = n > 0 && unsafe_load(buf.a, n) == 0x0d ? n - 1 : n # remove trailing '\r' if present

"""
    readuntil!(r::Bufio{T}, delim::Int = -1, offset = 0, keep::Bool = false) where T<:IO

Read data up to a delimitor `delim` into `r.bb` and return the number of bytes consumed.

# Arguments
- `r::Bufio{T}`: buffered stream opened for reading
- `delim::Int`: delimitor. -1 for line; -2 for space/TAB/line.
- `offset`: read data to r.bb + offset
- `keep::Bool`: whether to keep the delimiter

# Return
Number of bytes consumed. -1 for end-of-file. -2 for file reading error as is
returned from `Base.readbytes!()`.
"""
function readuntil!(r::Bufio{T}, delim::Int = -1, offset = 0, keep::Bool = false) where {T<:IO}
	if r.start > r.len && r.iseof return -1 end
	n = 0
	while true
		if r.start > r.len
			if r.iseof == false
				r.start, r.len = 1, Base.readbytes!(r.fp, r.buf)
				if r.len == 0 r.iseof = true; break end
				if r.len < 0 r.iseof = true; return -2 end
			else
				break
			end
		end
		x = r.len + 1
		if delim == -1 # use '\n' as the delimitor
			for i = r.start:r.len
				@inbounds if r.buf[i] == 0x0a x = i; break end
			end
		elseif delim == -2 # use ' ', '\t' or '\n' as the delimitor
			for i = r.start:r.len
				@inbounds if r.buf[i] == 0x20 || r.buf[i] == 0x09 || r.buf[i] == 0x0a x = i; break end
			end
		else
			for i = r.start:r.len
				@inbounds if r.buf[i] == delim x = i; break end
			end
		end
		l = keep && x <= r.len ? x - r.start + 1 : x - r.start
		reserve!(r.bb, UInt64(offset + n + l))
		unsafe_copyto!(r.bb.a + offset + n, pointer(r.buf, r.start), l)
		n += l
		r.start = x + 1
		if x <= r.len break end
	end
	if (delim == -1 || delim == -2) && !keep n = trimret(r.bb, n) end # remove trailing '\r' if present
	return n == 0 && r.iseof ? -1 : n
end

"""
    Base.readline(r::Bufio{T}) where T<:IO

Read a line and return it as a string on success or `nothing` on end-of-file.
"""
function Base.readline(r::Bufio{T}) where {T<:IO}
	n = readuntil!(r)
	n >= 0 ? tostring(r.bb, n) : nothing
end

#
# FastxReader
#
mutable struct FastxReader{T<:IO}
	r::Bufio{T}
	last::UInt8
	errno::Int

	FastxReader{T}(io::T) where {T<:IO} = new{T}(Bufio{T}(io), 0, 0)
end

"""
    FastxReader(io::T) where T<:IO

Wrap an IO stream and return a `FastxReader{T}` object.
"""
FastxReader(io::T) where {T<:IO} = FastxReader{T}(io)

mutable struct FastxRecord
	name::String
	seq::String
	qual::String
	comment::String
end

"""
    Base.read(f::FastxReader{T}) where T<:IO

Read a FASTA/FASTQ record and return a `FastxRecord` object on success or
`nothing` on end-of-file.

In the returned object `r`, `r.name` is the number of sequence; `r.comment` is
the comment line on the FASTA/Q header; `r.seq` is the sequnece; `r.qual` is
the quality. `r.comment` and `r.qual` may be empty strings if not available.
`f.errno` is 0 if no errors occurred during parsing, or a negative number on
errors.
"""
function Base.read(f::FastxReader{T}) where {T<:IO}
	if f.last == 0 # then jump to the next header line
		while (c = readbyte(f.r)) >= 0 && c != 0x3e && c != 0x40 end # 0x3e = '>', 0x40 = '@'
		if c < 0 return nothing end # normal exit: no records
		f.last = c
	end # else: the first header char has been read in the previous call
	name = comment = seq = "" # reset all members
	n = readuntil!(f.r, -2, 0, true)
	if n < 0 return nothing end # normal exit: EOF
	if unsafe_load(f.r.bb.a, n) == 0x0a # end-of-line; no comments
		n = trimret(f.r.bb, n - 1)
		name = tostring(f.r.bb, n)
	else # there are FASTX comments
		name = tostring(f.r.bb, n - 1)
		n = readuntil!(f.r)
		comment = tostring(f.r.bb, n)
	end
	ls = 0
	while (c = readbyte(f.r)) >= 0 && c != 0x3e && c != 0x40 && c != 0x2b # 0x2b = '+'
		if c == 0x0a continue end # skip empty lines
		reserve!(f.r.bb, UInt64(ls + 1))
		unsafe_store!(f.r.bb.a, c, ls + 1) # write the first character
		ls += 1
		ls += readuntil!(f.r, -1, ls) # read the rest of the line
	end
	if c == 0x3e || c == 0x40 f.last = c end # the first header char has been read
	seq = tostring(f.r.bb, ls) # sequence read
	@assert ls == lastindex(seq) # guard against UTF-8
	if c != 0x2b return FastxRecord(name, seq, "", comment) end # FASTA
	while (c = readbyte(f.r)) >= 0 && c != 0x0a end # skip the rest of '+' line
	if c < 0 f.errno = -3; return nothing end # error exit: no quality string
	lq = 0
	while lq < ls
		n = readuntil!(f.r, -1, lq)
		if n < 0 break end
		lq += n
	end
	f.last = 0
	if lq != ls f.errno = -2; return nothing end # error exit: qual string is of a different length
	qual = tostring(f.r.bb, lq) # quality read
	@assert lq == lastindex(qual) # guard against UTF-8
	return FastxRecord(name, seq, qual, comment)
end

end # module Klib

File: lib/klib.nim
import os, algorithm

const klibVer* = "0.1"

###################################
# Unix getopt() and getopt_long() #
###################################

proc getArgv*(): seq[string] =
  for i in 1 .. paramCount():
    result.add(paramStr(i))

iterator getopt*(argv: var seq[string], ostr: string, longopts: seq[
    string] = @[]): (string, string) =
  var
    pos = 0
    cur = 0
  while cur < argv.len:
    var
      lopt = "" # long option
      opt = '?' # short option
      arg = ""  # option argument
    while cur < argv.len: # look for an option
      if argv[cur][0] == '-' and argv[cur].len > 1: # an option or "--"
        if argv[cur].len == 2 and argv[cur][1] == '-': # "--"
          cur = argv.len
        break
      else: cur += 1
    if cur == argv.len: break
    let a = argv[cur]
    if a[0..1] == "--": # long option
      pos = -1
      var
        pos_eq = find(a, '=')
        o = a[2 ..< a.len]
        c = 0
        k = -1
        tmp = ""
      if pos_eq > 0:
        o = a[2 ..< pos_eq]
        arg = a[pos_eq+1 ..< a.len]
      for i, x in longopts: # look for matching long options
        var y = x
        if y[^1] == '=': y = y[0 ..< y.len-1]
        if o.len <= y.len and o == y[0 ..< o.len]: # prefix match
          k = i; c += 1; tmp = y
          if o == y: # exact option match
            c = 1
            break
      if c == 1: # find a unique match
        lopt = tmp
        if pos_eq < 0 and longopts[k][^1] == '=' and cur + 1 < argv.len:
          arg = argv[cur + 1]
          argv.delete(cur + 1)
    else: # short option
      if pos == 0: pos = 1
      opt = a[pos]
      pos += 1
      var k = find(ostr, opt)
      if k < 0: opt = '?'
      elif k + 1 < ostr.len and ostr[k + 1] == ':': # requiring an argument
        if pos >= a.len:
          if cur + 1 < argv.len:
            arg = argv[cur + 1]
            argv.delete(cur + 1)
        else:
          arg = a[pos ..< a.len]
        pos = -1
    if pos < 0 or pos >= argv[cur].len:
      argv.delete(cur)
      pos = 0
    if lopt != "": yield ("--" & lopt, arg)
    else: yield ("-" & opt, arg)

#################
# gzip file I/O #
#################

when defined(windows):
  const libz = "zlib1.dll"
elif defined(macosx):
  const libz = "libz.dylib"
  const libc = "libc.dylib"
else:
  const libz = "libz.so.1"
  const libc = "libc.so.6"

type
  gzFile = pointer

proc gzopen(path: cstring, mode: cstring): gzFile{.cdecl, dynlib: libz,
    importc: "gzopen".}
proc gzdopen(fd: int32, mode: cstring): gzFile{.cdecl, dynlib: libz,
    importc: "gzdopen".}
proc gzread(thefile: gzFile, buf: pointer, length: int): int32{.cdecl,
    dynlib: libz, importc: "gzread".}
proc gzclose(thefile: gzFile): int32{.cdecl, dynlib: libz, importc: "gzclose".}

type
  GzFile* = gzFile

proc open(f: var GzFile, fn: string,
    mode: FileMode = fmRead): int {.discardable.} =
  assert(mode == fmRead or mode == fmWrite)
  result = 0
  if fn == "-" or fn == "":
    if mode == fmRead: f = gzdopen(0, cstring("r"))
    elif mode == fmWrite: f = gzdopen(1, cstring("w"))
  else:
    if mode == fmRead: f = gzopen(cstring(fn), cstring("r"))
    elif mode == fmWrite: f = gzopen(cstring(fn), cstring("w"))
  if f == nil:
    result = -1
    raise newException(IOError, "error opening " & fn)

proc close(f: var GzFile): int {.discardable.} =
  if f != nil:
    result = int(gzclose(f))
    f = nil
  else: result = 0

proc read(f: var GzFile, buf: var string, sz: int, offset: int = 0):
    int {.discardable.} =
  if buf.len < offset + sz: buf.setLen(offset + sz)
  result = gzread(f, buf[offset].addr, buf.len)
  buf.setLen(result)

###################
# Buffered reader #
###################

type
  Bufio*[T] = tuple[fp: T, buf: string, st, en, sz: int, EOF: bool]

proc open*[T](f: var Bufio[T], fn: string, mode: FileMode = fmRead,
    sz: int = 0x10000): int {.discardable.} =
  assert(mode == fmRead) # only fmRead is supported for now
  result = f.fp.open(fn, mode)
  (f.st, f.en, f.sz, f.EOF) = (0, 0, sz, false)
  f.buf.setLen(sz)

proc xopen*[T](fn: string, mode: FileMode = fmRead,
    sz: int = 0x10000): Bufio[T] =
  var f: Bufio[T]
  f.open(fn, mode, sz)
  return f

proc close*[T](f: var Bufio[T]): int {.discardable.} =
  return f.fp.close()

proc eof*[T](f: Bufio[T]): bool {.noSideEffect.} =
  result = (f.EOF and f.st >= f.en)

proc readByte*[T](f: var Bufio[T]): int =
  if f.EOF and f.st >= f.en: return -1
  if f.st >= f.en:
    (f.st, f.en) = (0, f.fp.read(f.buf, f.sz))
    if f.en == 0: f.EOF = true; return -1
    if f.en < 0: f.EOF = true; return -2
  result = int(f.buf[f.st])
  f.st += 1

proc read*[T](f: var Bufio[T], buf: var string, sz: int,
    offset: int = 0): int {.discardable.} =
  if f.EOF and f.st >= f.en: return 0
  buf.setLen(offset)
  var off = offset
  var rest = sz
  while rest > f.en - f.st:
    if f.en > f.st:
      let l = f.en - f.st
      if buf.len < off + l: buf.setLen(off + l)
      copyMem(buf[off].addr, f.buf[f.st].addr, l)
      rest -= l
      off += l
    (f.st, f.en) = (0, f.fp.read(f.buf, f.sz))
    if f.en < f.sz: f.EOF = true
    if f.en == 0: return off - offset
  if buf.len < off + rest: buf.setLen(off + rest)
  copyMem(buf[off].addr, f.buf[f.st].addr, rest)
  f.st += rest
  return off + rest - offset

proc memchr(buf: pointer, c: cint, sz: csize_t): pointer {.cdecl, dynlib: libc,
    importc: "memchr".}

proc readUntil*[T](f: var Bufio[T], buf: var string, dret: var char,
    delim: int = -1, offset: int = 0): int {.discardable.} =
  if f.EOF and f.st >= f.en: return -1
  buf.setLen(offset)
  var off = offset
  var gotany = false
  while true:
    if f.en < 0: return -3
    if f.st >= f.en: # buffer is empty
      if not f.EOF:
        (f.st, f.en) = (0, f.fp.read(f.buf, f.sz))
        if f.en < f.sz: f.EOF = true
        if f.en == 0: break
        if f.en < 0:
          f.EOF = true
          return -2
      else: break
    var x: int = f.en
    if delim == -1: # read a line
      #for i in f.st..<f.en:
      #  if f.buf[i] == '\n': x = i; break
      var p = memchr(f.buf[f.st].addr, cint(0xa), csize_t(f.en - f.st))
      if p != nil: x = cast[int](p) - cast[int](f.buf[0].addr)
    elif delim == -2: # read a field
      for i in f.st..<f.en:
        if f.buf[i] == '\t' or f.buf[i] == ' ' or f.buf[i] == '\n':
          x = i; break
    else: # read to other delimitors
      #for i in f.st..<f.en:
      #  if f.buf[i] == char(delim): x = i; break
      var p = memchr(f.buf[f.st].addr, cint(delim), csize_t(f.en - f.st))
      if p != nil: x = cast[int](p) - cast[int](f.buf[0].addr)
    gotany = true
    if x > f.st: # something to write to buf[]
      let l = x - f.st
      if buf.len < off + l: buf.setLen(off + l)
      copyMem(buf[off].addr, f.buf[f.st].addr, l)
      off += l
    f.st = x + 1
    if x < f.en: dret = f.buf[x]; break
  if not gotany and f.eof(): return -1
  if delim == -1 and off > 0 and buf[off - 1] == '\r':
    off -= 1
    buf.setLen(off)
  return off - offset

proc readLine*[T](f: var Bufio[T], buf: var string): bool {.discardable.} =
  var dret: char
  var ret = readUntil(f, buf, dret)
  return if ret >= 0: true else: false

################
# Fastx Reader #
################

type
  FastxRecord* = tuple[seq, qual, name, comment: string, status, lastChar: int]

proc readFastx*[T](f: var Bufio[T], r: var FastxRecord): bool {.discardable.} =
  var x: int
  var c: char
  if r.lastChar == 0: # the header character hasn't been read yet
    while true:       # look for the header character '>' or '@'
      x = f.readByte()
      if x < 0 or x == int('>') or x == int('@'): break
    if x < 0: r.status = x; return false # end-of-file or stream error
    r.lastChar = x
  r.seq.setLen(0); r.qual.setLen(0); r.comment.setLen(0)
  x = f.readUntil(r.name, c, -2)
  if x < 0: r.status = x; return false       # EOF or stream error
  if c != '\n': f.readUntil(r.comment, c) # read FASTA/Q comment
  while true:         # read sequence
    x = f.readByte()  # read the first char on a line
    if x < 0 or x == int('>') or x == int('+') or x == int('@'): break
    if x == int('\n'): continue
    r.seq.add(char(x))
    f.readUntil(r.seq, c, -1, r.seq.len)  # read the rest of the seq line
  r.status = r.seq.len   # for normal records, this keeps the sequence length
  if x == int('>') or x == int('@'): r.lastChar = x
  if x != int('+'): return true
  while true:         # skip the rest of the "+" line
    x = f.readByte()
    if x < 0 or x == int('\n'): break
  if x < 0: r.status = x; return false  # error: no quality
  while true:         # read quality
    x = f.readUntil(r.qual, c, -1, r.qual.len)
    if x < 0 or r.qual.len >= r.seq.len: break
  if x == -3: r.status = -3; return false # other stream error
  r.lastChar = 0
  if r.seq.len != r.qual.len: r.status = -4; return false
  return true

#############
# Intervals #
#############

type
  Interval*[S,T] = tuple[st, en: S, data: T, max: S]

proc sort*[S,T](a: var seq[Interval[S,T]]) =
  a.sort do (x, y: Interval[S,T]) -> int:
    if x.st < y.st: -1
    elif x.st > y.st: 1
    else: 0

proc index*[S,T](a: var seq[Interval[S,T]]): int {.discardable.} =
  if a.len == 0: return 0
  var is_srt = true
  for i in 1..<a.len:
    if a[i-1].st > a[i].st:
      is_srt = false; break
  if not is_srt: a.sort()
  var last_i: int
  var last: S
  for i in countup(0, a.len-1, 2): # leaves (i.e. at level 0)
    (last_i, last, a[i].max) = (i, a[i].en, a[i].en)
  var k = 1
  while 1 shl k <= a.len: # process internal nodes in the bottom-up order
    let x = 1 shl (k - 1)
    let i0 = (x shl 1) - 1 # the first node at level k
    let step = x shl 2
    for i in countup(i0, a.len - 1, step): # traverse nodes at level k
      let el = a[i - x].max  # max value of the left child
      let er = if i + x < a.len: a[i + x].max else: last # of the right child
      var e = a[i].en
      if e < el: e = el
      if e < er: e = er
      a[i].max = e
    # point last_i to the parent of the original last_i
    last_i = if ((last_i shr k) and 1) != 0: last_i - x else: last_i + x
    if last_i < a.len and a[last_i].max > last: # update last accordingly
      last = a[last_i].max
    k += 1
  return k - 1

iterator overlap*[S,T](a: seq[Interval[S,T]], st: S, en: S): Interval[S,T] {.noSideEffect.} =
  var h: int = 0
  while 1 shl h <= a.len: h += 1
  h -= 1 # h is the height of the tree
  var stack: array[64, tuple[k, x, w:int]] # 64 is the max possible tree height
  var t: int = 0
  stack[t] = (h, (1 shl h) - 1, 0); t += 1 # push the root
  while t > 0: # the following guarantees sorted "yield"
    t -= 1
    let (k, x, w) = stack[t] # pop from the stack
    if k <= 3: # in a small subtree, traverse everything
      let i0 = (x shr k) shl k
      var i1 = i0 + (1 shl (k + 1)) - 1
      if i1 >= a.len: i1 = a.len
      for i in countup(i0, i1 - 1):
        if a[i].st >= en: break  # out of range; no need to proceed
        if st < a[i].en: yield a[i] # overlap! yield
    elif w == 0: # the left child not processed
      let y = x - (1 shl (k - 1)) # the left child of z.x; y may >=a.len
      stack[t] = (k, x, 1); t += 1
      if y >= a.len or a[y].max > st:
        stack[t] = (k-1, y, 0); t += 1 # add left child
    elif x < a.len and a[x].st < en: # need to push the right child
      if st < a[x].en: yield a[x] # test if x overlaps the query
      stack[t] = (k - 1, x + (1 shl (k - 1)), 0); t += 1
